File: app/src/main/java/su/linka/linkapaperboard/Cookie.java
package su.linka.linkapaperboard;

import android.content.Context;
import android.content.SharedPreferences;

class Cookie {
    private static final String STORAGE_NAME = "storage";
    private static final int DEF_VALUE = 3;
    private static final String GRID_SIZE_ID = "gridsize";
    private static final String GRID_UPPERCASE_ID = "griduppercase";

    private static Cookie instance;
    private static Context context;

    private final SharedPreferences mSharedPref;

    public static Cookie getInstance() {
        return getInstance(MainActivity.context);
    }

    private Cookie(Context cxt) {
        context = cxt;
        mSharedPref = context.getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE);

    }

    public int getGridSize() {
        return mSharedPref.getInt(GRID_SIZE_ID, DEF_VALUE);
    }

    public void setGridSize(int gridSize) {
        SharedPreferences.Editor editor = mSharedPref.edit();
        editor.putInt(GRID_SIZE_ID, gridSize);
        editor.apply();

    }

    public static Cookie getInstance(Context creationContext) {
        if (instance == null || context != creationContext) {
            instance = new Cookie(creationContext);
        }
        return instance;
    }

    public void setUppercaseChecked(boolean checked) {
        SharedPreferences.Editor editor = mSharedPref.edit();
        editor.putBoolean(GRID_UPPERCASE_ID, checked);
        editor.apply();

    }

    public boolean getUppercaseChecked() {
        return mSharedPref.getBoolean(GRID_UPPERCASE_ID, true);
    }
}

File: app/src/main/java/su/linka/linkapaperboard/Cookie.kt
package su.linka.linkapaperboard

import android.content.Context
import android.content.SharedPreferences
import androidx.annotation.VisibleForTesting

class Cookie private constructor(private val sharedPref: SharedPreferences) {

    val gridSize: Int
        get() = sharedPref.getInt(GRID_SIZE_ID, DEF_VALUE)

    fun setGridSize(gridSize: Int) {
        sharedPref.edit()
            .putInt(GRID_SIZE_ID, gridSize)
            .apply()
    }

    fun setUppercaseChecked(checked: Boolean) {
        sharedPref.edit()
            .putBoolean(GRID_UPPERCASE_ID, checked)
            .apply()
    }

    fun getUppercaseChecked(): Boolean = sharedPref.getBoolean(GRID_UPPERCASE_ID, true)

    val privacyNoticeAcknowledged: Boolean
        get() = sharedPref.getBoolean(PRIVACY_NOTICE_ACKNOWLEDGED_ID, false)

    fun acknowledgePrivacyNotice() {
        sharedPref.edit()
            .putBoolean(PRIVACY_NOTICE_ACKNOWLEDGED_ID, true)
            .apply()
    }

    companion object {
        private const val STORAGE_NAME = "storage"
        private const val DEF_VALUE = 3
        private const val GRID_SIZE_ID = "gridsize"
        private const val GRID_UPPERCASE_ID = "griduppercase"
        private const val PRIVACY_NOTICE_ACKNOWLEDGED_ID = "privacy_notice_acknowledged"

        fun from(context: Context): Cookie {
            val appContext = context.applicationContext
            val prefs = appContext.getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE)
            return Cookie(prefs)
        }

        @VisibleForTesting
        internal fun resetForTests(testContext: Context) {
            testContext.getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE)
                .edit()
                .clear()
                .commit()
        }
    }
}
